docs/Server/API/V1/relay.md

```diff
@@ -0,0 +1,143 @@
+structure; however with the `server.archiveonly` metadata key the Pbench Server
+can be used to archive and manage metadata for any tarball passed through a
+Relay server.
+
+## URI parameters
+
+`<uri>` string \
+The Relay server URI of the tarball's manifest `application/json` file. This
+must provide the following JSON keys:
+
+For example,
+
+```json
+{
+    "uri": "https://relay.example.com/52adfdd3dbf2a87ed6c1c41a1ce278290064b0455f585149b3dadbe5a0b62f44",
+    "md5": "22a4bc5748b920c6ce271eb68f08d91c".
+    "name": "fio_rw_2018.02.01T22.40.57.tar.xz".
+    "access": "private",
+    "metadata": ["server.origin:myrelay", "global.agent:cloud1"]
+}
+```
+
+## Manifest file keys
+
+`access`: [ `private` | `public` ] \
+The desired initial access scope of the dataset. Select `public` to make the
+dataset accessible to all clients, or `private` to make the dataset accessible
+only to the owner. The default access scope if the key is omitted from the
+manifest is `private`.
+
+For example, `"access": "public"`
+
+`md5`: tarball MD5 hash \
+The MD5 hash of the compressed tarball file. This must match the actual tarball
+octet stream specified by the manifest `uri` key.
+
+`metadata`: [metadata key/value strings] \
+A set of desired Pbench Server metadata key values to be assigned to the new
+dataset. You can set the initial resource name (`dataset.name`), for example, as
+well as assigning any keys in the `global` and `user` namespaces. See
+[metadata](../metadata.md) for more information.
+
+In particular the client can set any of:
+* `dataset.name`: [default dataset name](../metadata.md#datasetname)
+* `server.origin`: [dataset origin](../metadata.md#serverorigin)
+* `server.archiveonly`: [suppress indexing](../metadata.md#serverarchiveonly)
+* `server.deletion`: [default dataset expiration time](../metadata.md#serverdeletion).
+
+For example, `?metadata=server.archiveonly:true,global.project:oidc`
+
+`name`: The original tarball file name \
+The string value must represent a legal filename with the compound type of
+`.tar.xz` representing a `tar` archive compressed with the `xz` program.
+
+`uri`: Relay URI resolving to the tarball file \
+An HTTP `GET` on this URI, exactly as recorded, must return the original tarball
+file as an `application/octet-stream`.
+
+## Request headers
+
+`authorization: bearer` token \
+*Bearer* schema authorization assigns the ownership of the new dataset to the
+authenticated user. E.g., `authorization: bearer <token>`
+
+`content-length` tarball size \
+The size of the request octet stream in bytes. Generally supplied automatically by
+an upload agent such as Python `requests` or `curl`.
+
+## Response headers
+
+`content-type: application/json` \
+The return is a serialized JSON object with status information.
+
+## Response status
+
+`200`   **OK** \
+Successful request. The dataset MD5 hash is identical to that of a dataset
+previously uploaded to the Pbench Server. This is assumed to be an identical
+tarball.
```

Comment:
Does this response indicate that no `GET` request was or will be issued for the file listed in the manifest?  (Do we want to make a statement either way?)

I ask because it might be that the manifest `GET` request has a side effect (such as causing the file to be deleted or causing the relay to shut down), which the user may desire even if the result file has been previously uploaded.

Reply:
Yes, relay `GET` side-effects would complicate the semantics, and I'd really rather not have side-effects from any `GET`. Right now, `/api/v1/relay/<uri>` is, from the relay's point of view, a strictly read-only operation. We *could* `DELETE` either or both of the URIs to clean up, possibly even under client control with an `?delete` query parameter.

Reply:
OK, so we could, in principle, add a `?delete` query parameter to the `POST` API which would (conceivably) prompt it to issue `DELETE` requests on the same URIs as it issued (or would have issued, if it weren't a duplicate) `GET` requests on.  That would preserve the purity of the `GET` methods, and it would relieve us from having to worry about whether the Pbench Server actually makes all the `GET` calls that the client might have expected.

It does, however, mean that the client cannot freely use the Pbench Server as its agent in addressing the relay server.  That is, the Pbench Server will be assuming that it can do `DELETE` requests on the URIs that the client requested it do `GET` requests on, and it sort of implies that the client must exercise extreme care adding query parameters to the URLs that it hands to the Pbench Server or places into the manifest.

I was hoping that we could maintain a more black-box approach for the relay.  While I have no doubt that the file-relay thingie that I've built for us will be very handy and that it will integrate nicely with both the Pbench Agent and the Pbench Server, I was hoping that it would be straightforward for a client to instead use, e.g., AWS S3 as the relay, simply by constructing the appropriate manifest, uploading it to a bucket, and handing a pre-signed URL to the Pbench Server.  If the Pbench Server knows nothing other than to fetch the manifest and then fetch the URI embedded in it, then it doesn't need to know anything about the nature of the relay service.

And, if that's the case, then the client could add parameters to the URL for the manifest or to the URI in the manifest which could control the relay server (if that server was so inclined), which I think will prove handy, depending on the implementation of the server.

But, that presupposes that the Pbench Server will actually do all the `GET`'s....

I was considering a "fire and forget" scenario where the client starts the relay and walks away, knowing that, when the Pbench Server is done, the last request will cause the relay to clean up and shut down.  But, maybe that's not really practical or sufficiently robust.

Reply:
First, I think "black box" means not depending on nor coding around any assumptions that a `GET` is anything but the pure read-only operation one ought to be.

I'd certainly be open to a `POST /relay/<uri>?delete=true` that caused the Relay API to explicitly issue `DELETE` operations on both relay URIs upon successful completion. (Or do we also delete on failure on the theory that a retry is unlikely to have different results? Maybe only if it's an explicit server failure rather than a connection failure? Etc...)

Syntactically, "adding query parameters" to the Relay manifest URI is problematic; we'd need to be really careful about quoting (if it's even possible) to prevent Flask from parsing them as *our* query parameters. To generalize the relay implementation we'd likely also need to worry about authentication mechanisms.

I think this is all "one plus" stuff we can deal with later.

Reply:
I think we should separate the Pbench Server API from the file relay's API.  (Thus, the relative merits of addressing the Pbench Server via `GET` vs. `POST` are an interesting but orthogonal discussion.)  What I'm trying to hold out for is the proper abstraction of the file relay.

I think that the only requirement should be that the Pbench Server will do a `GET` on whatever URL the User provides, and that the `GET` will produce a properly formatted manifest file contents.  We can _expect_ that the URL is produced by the Pbench Agent on behalf of the User, but I don't think that we should require it.  Likewise, we can _expect_ that the URI in the manifest was produced by the Pbench Agent based on information from the User, but I don't think that we should require that, either.  I don't see any reason to presuppose that some other client won't be involved.

Given that, I don't think we should make any assumptions about there being no query parameters present in the URLs or the URIs, and I don't think we should be making any stipulations about side-effects of the `GET` requests on them.  Those are the interface contract between the User and the relay, and they are outside of the contract between the User/client and the Pbench Server.

So, yes, this is stuff to be dealt with later, so long as we don't foreclose on any of it now.  And, being clear that, if the Pbench Server returns a `200` here, the secondary `GET` request will not have been made, goes some way toward supporting that.